src/main/java/net/snowflake/client/core/SFBaseSession.java

```diff
@@ -341,42 +341,68 @@ public HttpClientSettingsKey getHttpClientKey() throws SnowflakeSQLException {
+            "http.useProxy={}, http.proxyHost={}, http.proxyPort={}, https.proxyHost={}, https.proxyPort={}, http.nonProxyHosts={}, NO_PROXY={}",
             httpUseProxy,
             httpProxyHost,
             httpProxyPort,
             httpsProxyHost,
             httpsProxyPort,
+            nonProxyHosts,
             noProxy);
-      } else {
-        logger.debug("http.useProxy={}. JVM proxy not used.", httpUseProxy);
-      }
-      // Set a new HttpClientSettingsKey with the https params. If the proxy is http, the native
-      // Java implementation will work.
-      if (httpUseProxy
-          && !Strings.isNullOrEmpty(httpsProxyHost)
-          && !Strings.isNullOrEmpty(httpsProxyPort)) {
-        int proxyPort;
-        try {
-          proxyPort = Integer.parseInt(httpsProxyPort);
-        } catch (NumberFormatException | NullPointerException e) {
-          throw new SnowflakeSQLException(
-              ErrorCode.INVALID_PROXY_PROPERTIES, "Could not parse port number");
+        // There are 2 possible parameters for non proxy hosts that can be combined into 1
+        String combinedNonProxyHosts = Strings.isNullOrEmpty(nonProxyHosts) ? "" : nonProxyHosts;
+        if (!Strings.isNullOrEmpty(noProxy)) {
+          combinedNonProxyHosts += combinedNonProxyHosts.length() == 0 ? "" : "|";
+          combinedNonProxyHosts += noProxy;
+        }
+        if (!Strings.isNullOrEmpty(httpsProxyHost) && !Strings.isNullOrEmpty(httpsProxyPort)) {
+          int proxyPort;
+          try {
+            proxyPort = Integer.parseInt(httpsProxyPort);
+          } catch (NumberFormatException | NullPointerException e) {
+            throw new SnowflakeSQLException(
+                ErrorCode.INVALID_PROXY_PROPERTIES, "Could not parse port number");
+          }
+          ocspAndProxyKey =
+              new HttpClientSettingsKey(
+                  getOCSPMode(),
+                  httpsProxyHost,
+                  proxyPort,
+                  combinedNonProxyHosts,
+                  "", /* user = empty */
+                  "", /* password = empty */
+                  "https");
+        } else if (!Strings.isNullOrEmpty(httpProxyHost) && !Strings.isNullOrEmpty(httpProxyPort)) {
+          int proxyPort;
+          try {
+            proxyPort = Integer.parseInt(httpProxyPort);
+          } catch (NumberFormatException | NullPointerException e) {
+            throw new SnowflakeSQLException(
+                ErrorCode.INVALID_PROXY_PROPERTIES, "Could not parse port number");
+          }
+          ocspAndProxyKey =
+              new HttpClientSettingsKey(
+                  getOCSPMode(),
+                  httpProxyHost,
+                  proxyPort,
+                  combinedNonProxyHosts,
+                  "", /* user = empty */
+                  "", /* password = empty */
+                  "http");
+        } else {
```

Comment:
Case where useProxy=true but no other valid proxy settings are there.

Reply:
I have one suggestion about this part of code. The nested if-else is not quite easy to read. Basically, here there are only three branches:

```
if (httpUseProxy
          && !Strings.isNullOrEmpty(httpsProxyHost)
          && !Strings.isNullOrEmpty(httpsProxyPort)) {
  ...
  return new HttpClientSettingsKey(
                  getOCSPMode(),
                  httpsProxyHost,
                  proxyPort,
                  combinedNonProxyHosts,
                  "", /* user = empty */
                  "", /* password = empty */
                  "https");
}

if (httpUseProxy
          && !Strings.isNullOrEmpty(httpProxyHost) 
          && !Strings.isNullOrEmpty(httpProxyPort)) {
  ...
  return new HttpClientSettingsKey(
                  getOCSPMode(),
                  httpProxyHost,
                  proxyPort,
                  combinedNonProxyHosts,
                  "", /* user = empty */
                  "", /* password = empty */
                  "http");
}

return new HttpClientSettingsKey(getOCSPMode());
```

Reply:
I'd like to keep it as is to reuse as much of the code as possible. With the current setup, the combinedNonProxyHost setting can be reused across all 3 if/else cases. If I un-nested it, I'd have to redo this logic in each branch.

Reply:
I don't get it. Why cannot reuse `combinedNonProxyHost` after unnesting it?
```
String combinedNonProxyHosts = "";
if (httpUseProxy) { udpate combinedNonProxyHosts; }

if (httpUseProxy
          && !Strings.isNullOrEmpty(httpsProxyHost)
          && !Strings.isNullOrEmpty(httpsProxyPort)) {
  ...
  return new HttpClientSettingsKey(
                  getOCSPMode(),
                  httpsProxyHost,
                  proxyPort,
                  combinedNonProxyHosts,
                  "", /* user = empty */
                  "", /* password = empty */
                  "https");
}

if (httpUseProxy
          && !Strings.isNullOrEmpty(httpProxyHost) 
          && !Strings.isNullOrEmpty(httpProxyPort)) {
  ...
  return new HttpClientSettingsKey(
                  getOCSPMode(),
                  httpProxyHost,
                  proxyPort,
                  combinedNonProxyHosts,
                  "", /* user = empty */
                  "", /* password = empty */
                  "http");
}

return new HttpClientSettingsKey(getOCSPMode());
```
Another option is 
```
if (httpUseProxy) { 
  udpate combinedNonProxyHosts; 

  if (!Strings.isNullOrEmpty(httpsProxyHost)
          && !Strings.isNullOrEmpty(httpsProxyPort)) {
  ...
  return new HttpClientSettingsKey(
                  getOCSPMode(),
                  httpsProxyHost,
                  proxyPort,
                  combinedNonProxyHosts,
                  "", /* user = empty */
                  "", /* password = empty */
                  "https");
  }
  
  if (!Strings.isNullOrEmpty(httpProxyHost) 
            && !Strings.isNullOrEmpty(httpProxyPort)) {
    ...
    return new HttpClientSettingsKey(
                    getOCSPMode(),
                    httpProxyHost,
                    proxyPort,
                    combinedNonProxyHosts,
                    "", /* user = empty */
                    "", /* password = empty */
                    "http");
  }
}

return new HttpClientSettingsKey(getOCSPMode());
```